bitwarden_license/src/Commercial.Core/SecretsManager/Commands/Porting/ImportCommand.cs

```diff
@@ -0,0 +1,110 @@
+    }
+
+    public async Task<SMImport> ImportAsync(Guid organizationId, SMImport import)
+    {
+        var importedProjects = new List<Project>();
+        var importedSecrets = new List<Secret>();
+
+        try
+        {
+            import = AssignNewIds(import);
+
+            if (import.Projects != null && import.Projects.Any())
+            {
+                importedProjects = (await _projectRepository.ImportAsync(import.Projects.Select(p => new Project
+                {
+                    Id = p.Id,
+                    OrganizationId = organizationId,
+                    Name = p.Name,
+                }))).ToList();
+            }
+
+            if (import.Secrets != null && import.Secrets.Any())
+            {
+                importedSecrets = (await _secretRepository.ImportAsync(import.Secrets.Select(s => new Secret
+                {
+                    Id = s.Id,
+                    OrganizationId = organizationId,
+                    Key = s.Key,
+                    Value = s.Value,
+                    Note = s.Note,
+                    Projects = s.ProjectIds != null && s.ProjectIds.Any() ? s.ProjectIds.Select(id => new Project { Id = id }).ToList() : null,
+                }))).ToList();
+            }
+        }
+        catch (Exception)
+        {
+            if (importedProjects.Any())
+            {
+                // remove them
+            }
+
+            if (importedSecrets.Any())
+            {
+                // remove them
+            }
+
+            throw new Exception("Error attempting import");
+        }
+
+        return import;
+    }
+
+    public SMImport AssignNewIds(SMImport import)
+    {
+        Dictionary<Guid, Guid> oldNewProjectIds = new Dictionary<Guid, Guid>();
+        var projects = new List<SMImport.InnerProject>();
+        var secrets = new List<SMImport.InnerSecret>();
+
+        if (import.Projects != null && import.Projects.Any())
+        {
+            foreach (var project in import.Projects)
+            {
+                var newProjectId = Guid.NewGuid();
+                oldNewProjectIds.Add(project.Id, newProjectId);
+                projects.Add(new SMImport.InnerProject
+                {
+                    Id = newProjectId,
+                    Name = project.Name,
+                });
+            }
+        }
+
+        if (import.Secrets != null && import.Secrets.Any())
+        {
+            foreach (var secret in import.Secrets)
+            {
+                secrets.Add(new SMImport.InnerSecret
+                {
+                    Id = Guid.NewGuid(),
+                    Key = secret.Key,
+                    Value = secret.Value,
+                    Note = secret.Note,
+                    ProjectIds = secret.ProjectIds?.Select(projectId => oldNewProjectIds[projectId]),
+                });
+            }
+        }
+
+        return new SMImport
+        {
+            Projects = projects,
+            Secrets = secrets
+        };
```

Comment:
I believe this should be equivalent.

Changed to use `ToDictionary` instead of foreach loop. We can skip the lookup table by using a dictionary directly which we then convert back to a list.

```suggestion
        var projects = new Dictionary<Guid, SMImport.InnerProject>();
        var secrets = new List<SMImport.InnerSecret>();

        if (import.Projects?.Any() == true)
        {
            projects = import.Projects.ToDictionary(
                p => p.Id,
                p => new SMImport.InnerProject { Id = Guid.NewGuid(), Name = p.Name });
        }

        if (import.Secrets?.Any() == true)
        {
            foreach (var secret in import.Secrets)
            {
                secrets.Add(new SMImport.InnerSecret
                {
                    Id = Guid.NewGuid(),
                    Key = secret.Key,
                    Value = secret.Value,
                    Note = secret.Note,
                    ProjectIds = secret.ProjectIds?.Select(id => projects[id].Id),
                });
            }
        }

        return new SMImport
        {
            Projects = projects.Values,
            Secrets = secrets,
        };
```

Reply:
Thanks for this recommendation. I implemented this change, I think your solution is cleaner. However, I'm just curious - isn't the performance very similar? In this new code, you still have to loop through to create the dictionary (line 104), and when adding to a dictionary, the runtime still has to check if that id exists or not. Then, you still have the `Select` statement for the secrets in either solution. The only main difference is in the old solution we also added to the projects list, but I believe that is still an O(1) operation (worst case).

Reply:
You are right, must have mixed it up slightly (a good reason for not reviewing a lot in a row 😢. I thought we iterated all the projects to extract it, but I see that we used a lookup table for it.